feat(tree): Implement TreeView v2 API (#20815)
microsoft/FluidFramework: packages/dds/tree/docs/user-facing/schema-evolution.md

> # Schema Evolution

Reviewer: These docs are great! Thank you!!

This document outlines the general constraints around maintaining compatibility in an application document ecosystem in the face of schema changes.

## Context

When an application creates and uses a schema to open a document using `SchemaFactory` and `TreeView.initialize`,
information about that schema is persisted as part of the document.
The schema that the application works with in code is hereforth referred to as "view schema" and the schema persisted with the document "stored schema."

> Compatibility scenarios can be described in terms of discrepancies between the view schema and the stored schema.

Reviewer 2: Are we still planning to support something like differences to track the actual mismatches between schemas, similar to the contents within `treeDiscrepancies.ts` in [v2](https://github.com/microsoft/FluidFramework/blob/b2f9c65b29d555c0b299356ba004d35fa220cb14/packages/dds/tree/src/simple-tree/treeDiscrepancies.ts)? It seems that the definition of the discrepancy has changed.

	y: builder.number,
}) {}

class Circle extends builder.object({
	center: Point,
	radius: builder.number,
}) {}

class Plane extends builder.object({
	shapes: builder.array([Circle, Point]),
}) {}
```

but discover that the document's stored schema doesn't allow `Point` children under the `shapes` field, i.e. its definition of `Plane` was this:

```typescript
class Plane extends builder.object({
	shapes: builder.array([Circle]),
}) {}
```

Applications may leverage `SharedTree`'s compatibility APIs to implement policies that allow clients running incompatible code to fail in predictable ways.

## General Approach

SharedTree allows application authors to make _backward-compatible changes to document schema_ using `TreeView.upgradeSchema`.
A change to document schema is backward-compatible if the set of documents allowed by the old schema is a subset of documents allowed by the new schema.
Put differently, valid documents must remain valid.

The following types of changes are backward-compatible under this definition:

1. Adding an optional field to an object node
2. Adding to the set of allowed types for a field
3. Relaxing a field kind to a more general field kind

> Adding new schemas (rather than modifying existing ones) is allowed and a typical reason to make a change to the allowed types for a field.

Reviewer: Nit (I had to reread in order to parse correctly)
```suggestion
Adding new schemas (rather than modifying existing ones) is allowed and is a typical reason to make a change to the allowed types for a field.
```

Examples for each type of change can be found below.
Each change should be interpreted independently from the others (rather than in aggregate):

```typescript
// Base schema:

> const builder = new SchemaFactory("Geometry");

Reviewer: Nit: for user-facing examples, we should probably come up with a consistent naming convention for `SchemaFactory` variables. We have some pretty wide disagreement in naming across our documents and examples.

Probably not a concern for this PR, but something for the team to consider.

class Point extends builder.object({
	x: builder.number,
	y: builder.number,
}) { }

class Circle extends builder.object({
	center: Point,
	radius: builder.number
}) { }

class Plane extends builder.object({
	shapes: builder.array([Circle, Point])
}) { }

// ---------------------------------------------------------
// Change type 1: adding an optional field to an object node

// Suppose the declaration of `Circle` was updated to:
class Circle extends builder.object({
	center: Point,
	radius: builder.number,
	color: builder.optional(builder.string),
}) { }

// -------------------------------------------------------------
// Change type 2: adding to the set of allowed types for a field

> // Suppose the application wants to add support for a "square" notion like this:

Reviewer: This sort of change is backwards compatible with existing _documents_, but not necessarily with existing view schemas. The other 2 classes of changes _are_ compatible with existing view schemas, so it might be worth calling this difference out somewhere?

class Square extends builder.object({
	length: builder.number,
}} { }
// They then update their "Plane" definition to allow it as a valid child:
class Plane extends builder.object({
	shapes: builder.array([Circle, Point, Square])
}) { }

// ------------------------------------
// Change type 3: relaxing a field kind

> // Imagine that the application has added code allowing circles to have a default radius of 1.

Reviewer 3: Since defaults are a feature we are landing which defaults a value when creating a node, not when reading it, this sentence could be better future proofed and clarified with a little rephrasing.

```suggestion
// Imagine that the application has added code allowing circles have a radius inherited from a some setting above it in the tree if not explicitly specified:
```

Author: The scenario also applies when an application has hardcoded default values completely external to tree code, no? I'd rather use a phrasing which isn't specific to some defaulting mechanism.

			return;
		}

		// Schema changes invalidate the root of the tree. Real applications would want to react to that here.
	});

	// Application can open document!
	renderApplication(view.root);
}
```

One consequence of this approach in isolation is that clients on old code versions can be "locked out" of documents when newer clients upgrade them,
at least until the clients running old code are able to upgrade to the newer application version.
This might be acceptable for applications with very quick deployment cadences.

To mitigate this in ecosystems with larger version skew, application authors could roll out schema changes by first deploying an application version which understands both the newer and older schema, but avoids upgrading to the newer schema.

Once the application version that understands both schemas saturates sufficiently, the application can start upgrading documents to the newer schema.

#### Allow optional field additions

> **_WARNING:_** The policy outlined in this section is not currently implementable. There are plans to extend the `compatibility` API with information that allows implementing policies such as this, but the exact API is not finalized.

> Adding an optional field ot an object node is one of the more safe types of schema changes from the perspective of clients running older code collaborating with clients running newer code.

Reviewer: ```suggestion
Adding an optional field to an object node is one of the safer types of schema changes from the perspective of clients running older code collaborating with clients running newer code.
```

This is because newer code must already have fallback behavior for absence of the optional field for backward-compatibility reasons, and older application code can generally just "ignore the extra field"

> That isn't strictly true--e.g. old code code using a spread operation or reflection APIs like `Object.keys` might fail upon encountering objects with a populated optional field it wasn't expecting--but this caveat might be acceptable for application authors enough for them to allow collaboration between such clients.

Reviewer: ```suggestion
That isn't strictly true--e.g. old client code using a spread operation or reflection APIs like `Object.keys` might fail upon encountering objects with a populated optional field it wasn't expecting--but this caveat might be acceptable for application authors enough for them to allow collaboration between such clients.
```
